ecl/hqlcpp/hqlttcpp.cpp

```diff
@@ -8865,6 +8865,226 @@ IHqlExpression * FilteredIndexOptimizer::createTransformed(IHqlExpression * expr
 //==============================================================================================================
 
 
+static HqlTransformerInfo DFSLayoutTransformerInfo("DFSLayoutTransformer");
+DFSLayoutTransformer::DFSLayoutTransformer(IErrorReceiver &_errs, ICodegenContextCallback * _ctxCallback, HqlCppOptions const &_options)
+: NewHqlTransformer(DFSLayoutTransformerInfo), errs(_errs), options(_options)
+{
+    ctxCallback = _ctxCallback;
+}
+
+
+IHqlExpression * DFSLayoutTransformer::createTransformed(IHqlExpression * expr)
+{
+    OwnedHqlExpr transformed = NewHqlTransformer::createTransformed(expr);
+    int filenameIdx = 0;
+    int recordIdx = 0;
+    switch (transformed->getOperator())
+    {
+    case no_table:
+        filenameIdx = 0;
+        recordIdx = 1;
+        break;
+    case no_keyindex:
+        throwUnexpected();  // They have already been translated into no_newkeyindex by HqlTreeNormalizer::transformKeyIndex
+        break;
+    case no_newkeyindex:
+        // Args are flatfile, record, transform, name. Offset of first payload field indicated by value of __payload_Atom attribute
+        filenameIdx = 3;
+        recordIdx = 1;
+        break;
+    default:
+        return transformed.getClear();
+    }
+    LinkedHqlExpr ds = transformed;
+    IHqlExpression * filename = ds->queryChild(filenameIdx);
+    bool translate = getBoolAttribute(ds, lookupAtom, options.translateDFSlayouts && filename->isConstant());
+    if (translate)
+    {
+        ECLlocation where(ds);
+        OwnedHqlExpr folded  = foldHqlExpression(errs, filename, nullptr, HFOforcefold);
+        if (!folded->isConstant())
```

Comment:
potentially not a strong enough check.  May need a new function isConstantValue(IHqlExpression *)??